frontend/src/data/saga.js

```diff
@@ -25,14 +26,40 @@ function* watchSingleRequest() {
 
 function* watchAddUser(action) {
   let { name } = action.payload;
+  yield performAsyncOperation(() => firebaseInteractor.addUser(name), "id");
+}
+
+function* watchDownloadImage(action) {
+  yield performAsyncOperation(
+    () => firebaseInteractor.downloadImage(action.payload.imageURL),
+    "imageURL"
+  );
+}
+
+/**
+ * Turns an async operation into a generator
+ * This at the end will put the `success` action into the value which will send the data through saga
+ *
+ * @param {Function<Void, Promise<T>>} operation the async operation to perform
+ * @param {String} payloadName the name of the object in the payload
+ * @template T
+ * Example:
+ * yield performAsyncOperation(() => new Promise((resolve, reject) => resolve(1)), "name")
+ * will generate a put(success({ name: 1 }))
+ */
+function* performAsyncOperation(operation, payloadName) {
+  let value;
+  const updateWithSuccess = async () => {
+    value = await operation();
+  };
   try {
-    let id;
-    const updateWithSuccess = async () => {
-      id = await firebaseInteractor.addUser(name);
-    };
     yield call(updateWithSuccess);
```

Comment:
I wonder if you can just do yield call(operation) because yield call is already doing asynchronous call

Reply:
The reason I added the `updateWithSuccess` function is that I need to actually get the value from the response to send to the success function.

Reply:
ah gotcha